### ISSUE TITLE
Trailing Slash on Route

### ISSUE THREAD

**Issue by chibicode** [hanami/hanami.github.io#118]
I'm doing the getting started tutorial:

http://lotusrb.org/guides/getting-started/

But I'm stuck here:

![image](https://cloud.githubusercontent.com/assets/992008/12320828/a6191e6c-ba5f-11e5-8238-b9d4063ffa18.png)

It seems that if there's no trailing slash on `/books`, it says "File not found":

![image](https://cloud.githubusercontent.com/assets/992008/12320835/bb7f86d8-ba5f-11e5-8b70-6ac6f0475122.png)

By adding the trailing slash, it displays correctly:

![image](https://cloud.githubusercontent.com/assets/992008/12320845/c5eb0af2-ba5f-11e5-90b6-69cabc18cd47.png)

So I had to change `visit` to use trailing slash in order to make the tests pass.

![image](https://cloud.githubusercontent.com/assets/992008/12320861/dd5ca1dc-ba5f-11e5-925b-b461343a9537.png)

Any ideas why this is happening?

**Comment by jodosha** [comment]
@chibicode Hey, I'm sorry about this trouble. Can you please paste here your routes? Thank you!

**Comment by chibicode** [comment]
@jodosha, thanks for a quick reply! Should be the same as the one on the tutorial...

```rb
get '/books', to: 'books#index'
get '/', to: 'home#index'
```

**Comment by chibicode** [comment]
Here's the repo (run `bundle install` and `rake test` to see the failure)
https://github.com/chibicode/lotus-bookshelf

**Comment by jodosha** [comment]
@chibicode I did exactly what you've asked, but didn't got any failure.

```shell
➜ bundle exec rake
Run options: --seed 5405

# Running:

....

Finished in 0.017537s, 228.0925 runs/s, 285.1157 assertions/s.

4 runs, 5 assertions, 0 failures, 0 errors, 0 skips
```

---

Can you please paste here the stack trace? Thank you!

**Comment by chibicode** [comment]
hmm sorry I can't figure out why... :(

```
❯ rake test
Run options: --seed 11830

# Running:

E...

Finished in 0.020582s, 194.3465 runs/s, 194.3465 assertions/s.

  1) Error:
List books#test_0001_displays each book on the page:
Capybara::ElementNotFound: Unable to find css "#books"
    /Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.5.0/lib/capybara/node/finders.rb:43:in `block in find'
    /Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.5.0/lib/capybara/node/base.rb:84:in `synchronize'
    /Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.5.0/lib/capybara/node/finders.rb:32:in `find'
    /Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.5.0/lib/capybara/session.rb:698:in `block (2 levels) in <class:Session>'
    /Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.5.0/lib/capybara/session.rb:287:in `within'
    /Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/capybara-2.5.0/lib/capybara/dsl.rb:51:in `block (2 levels) in <module:DSL>'
    /Users/shu/code/_cloned/bookshelf/spec/web/features/list_books_spec.rb:7:in `block (2 levels) in <top (required)>'

4 runs, 4 assertions, 0 failures, 1 errors, 0 skips
rake aborted!
Command failed with status (1): [ruby -I"lib:spec" -I"/Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rake-10.5.0/lib" "/Users/shu/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/rake-10.5.0/lib/rake/rake_test_loader.rb" "spec/**/*_spec.rb" ]

Tasks: TOP => test
(See full trace by running task with --trace)
```

also:

```
❯ ruby --version
ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-darwin14]
```

also:

```
❯ cat Gemfile.lock
GEM
  remote: https://rubygems.org/
  specs:
    capybara (2.5.0)
      mime-types (>= 1.16)
      nokogiri (>= 1.3.3)
      rack (>= 1.0.0)
      rack-test (>= 0.5.4)
      xpath (~> 2.0)
    dotenv (2.1.0)
    http_router (0.11.2)
      rack (>= 1.0.0)
      url_mount (~> 0.2.1)
    lotus-assets (0.1.0)
      lotus-helpers (~> 0.2, >= 0.2.5)
      lotus-utils (~> 0.6)
      tilt (~> 2.0, >= 2.0.2)
    lotus-controller (0.5.0)
      lotus-utils (~> 0.6)
      lotus-validations (~> 0.4)
      rack (~> 1.6, >= 1.6.2)
    lotus-helpers (0.2.6)
      lotus-utils (~> 0.6)
    lotus-mailer (0.1.0)
      lotus-utils (~> 0.5)
      mail (~> 2.5)
      tilt (~> 2.0, >= 2.0.1)
    lotus-model (0.5.1)
      lotus-utils (~> 0.6)
      sequel (~> 4.9)
    lotus-router (0.5.0)
      http_router (~> 0.11)
      lotus-utils (~> 0.6)
    lotus-utils (0.6.0)
    lotus-validations (0.4.0)
      lotus-utils (~> 0.6)
    lotus-view (0.5.0)
      lotus-utils (~> 0.6)
      tilt (~> 2.0, >= 2.0.1)
    lotusrb (0.6.0)
      bundler (~> 1.6)
      dotenv (~> 2.0)
      lotus-assets (~> 0.1)
      lotus-controller (~> 0.5)
      lotus-helpers (~> 0.2, >= 0.2.6)
      lotus-mailer (~> 0.1)
      lotus-router (~> 0.5)
      lotus-utils (~> 0.6)
      lotus-view (~> 0.5)
      shotgun (~> 0.9)
      thor (~> 0.19)
    mail (2.6.3)
      mime-types (>= 1.16, < 3)
    mime-types (2.99)
    mini_portile2 (2.0.0)
    minitest (5.8.3)
    nokogiri (1.6.7.1)
      mini_portile2 (~> 2.0.0.rc2)
    pg (0.18.4)
    rack (1.6.4)
    rack-test (0.6.3)
      rack (>= 1.0)
    rake (10.5.0)
    sequel (4.30.0)
    shotgun (0.9.1)
      rack (>= 1.0)
    thor (0.19.1)
    tilt (2.0.2)
    url_mount (0.2.1)
      rack
    xpath (2.0.0)
      nokogiri (~> 1.3)

PLATFORMS
  ruby

DEPENDENCIES
  bundler
  capybara
  lotus-model (~> 0.5)
  lotusrb (= 0.6.0)
  minitest
  pg
  rake
```

**Comment by jodosha** [comment]
@chibicode Our `Gemfile.lock` files are identical:

```shell
➜ diff Gemfile.lock GGemfile.lock
84a85
>
```

**Comment by jodosha** [comment]
@chibicode However, I found something really interesting:

```ruby
# spec/web/features/list_books_spec.rb
require 'features_helper'

describe 'List books' do
  it 'displays each book on the page' do
    visit '/books'

    # NEW ASSERTION
    current_path.must_equal '/books'

    within '#books' do
      assert page.has_css?('.book', count: 2), "Expected to find 2 books"
    end
  end
end
```

If I run again the tests, I've got an odd failure:

```shell
➜ bundle exec rake
Run options: --seed 37441

# Running:

..F.

Finished in 0.016184s, 247.1580 runs/s, 308.9475 assertions/s.

  1) Failure:
List books#test_0001_displays each book on the page [/Users/luca/Desktop/lotus-bookshelf/spec/web/features/list_books_spec.rb:7]:
Expected: "/books"
  Actual: "//books"

4 runs, 5 assertions, 1 failures, 0 errors, 0 skips
rake aborted!
Command failed with status (1): [ruby -I"lib:spec" -I"/Users/luca/.gem/ruby/2.3.0/gems/rake-10.5.0/lib" "/Users/luca/.gem/ruby/2.3.0/gems/rake-10.5.0/lib/rake/rake_test_loader.rb" "spec/**/*_spec.rb" ]

Tasks: TOP => default => test
(See full trace by running task with --trace)
```

Can you please try yourself?

**Comment by chibicode** [comment]
@jodosha I just edited and added the stack trace for `bundle exec lotus server` on the comment above just in case you didn't see it.

The strange thing is, if I change the route to something like

```
get '/foo', to: 'books#index'
```

Then `/foo` (without trailing slash) works fine.

```
❯ cat apps/web/config/routes.rb

get '/foo', to: 'books#index'
# Configure your routes here
# See: http://www.rubydoc.info/gems/lotus-router/#Usage
get '/', to: 'home#index'

❯ curl -I http://localhost:2300/foo
HTTP/1.1 200 OK
Server: WEBrick/1.3.1 (Ruby/2.2.0/2014-12-25)
Date: Fri, 15 Jan 2016 09:00:42 GMT
Content-Length: 0
Connection: Keep-Alive
```

**Comment by chibicode** [comment]
> Can you please try yourself?

Sure let me take a look.

**Comment by chibicode** [comment]
For me, `current_path` actually works and the error is on the same line as before. I commented out the last assertion:

![image](https://cloud.githubusercontent.com/assets/992008/12349345/d600dccc-bb23-11e5-9557-bda6e9b77016.png)

Then the test passes:

![image](https://cloud.githubusercontent.com/assets/992008/12349349/db613f68-bb23-11e5-892c-84e1acee34a5.png)

**Comment by chibicode** [comment]
I tried puts'ing `page.body`

![image](https://cloud.githubusercontent.com/assets/992008/12349391/3698bcf8-bb24-11e5-879c-4971fc054582.png)

**Comment by chibicode** [comment]
Let me know if there are other ways I might be able to debug the request!

**Comment by chibicode** [comment]
With trailing slash:

![image](https://cloud.githubusercontent.com/assets/992008/12349438/8e4d5170-bb24-11e5-9178-6af97afb08d6.png)

**Comment by CassioGodinho** [comment]
I also ran into this issue.
```ruby
visit '/books/'
current_path.must_equal '/books/'
```
Results in:
```
Minitest::Assertion: Expected: "/books/"
Actual: "//books/"
```

**Comment by joneslee85** [comment]
We have identified the issue and now working a patch. We will update on the status asap.

**Comment by joneslee85** [comment]
It has been addressed in ver 0.7.1

**Comment by jodosha** [comment]
:+1:
